pkg/processor/procbuiltin/unwrap.go

```diff
@@ -0,0 +1,310 @@
+	}
+
+	// get payload
+	var structPayload record.StructuredData
+	payload := structData["payload"] // the payload has the debezium record
+	if p, ok := payload.(map[string]interface{}); ok {
+		structPayload = p
+	} else {
+		return record.Record{}, cerrors.Errorf("payload doesn't contain a record")
+	}
+
+	// check fields under payload
+	err := d.validateRecord(structPayload)
+	if err != nil {
+		return record.Record{}, err
+	}
+
+	var before record.StructuredData
+	if b, ok := structPayload[DebeziumFieldBefore].(map[string]any); ok {
+		before = b
+	}
+
+	var after record.StructuredData
+	if a, ok := structPayload[DebeziumFieldAfter].(map[string]any); ok {
+		after = a
+	}
+
+	var op string
+	if str, ok := structPayload[DebeziumFieldOp].(string); ok {
+		op = str
+	} else {
+		return record.Record{}, cerrors.Errorf("%s operation is not a string", op)
+	}
+
+	operation, err := d.convertOperation(op)
+	if err != nil {
+		return record.Record{}, cerrors.Errorf("error unwrapping operation: %w", err)
+	}
+
+	metadata, err := d.unwrapMetadata(rec, structPayload)
+	if err != nil {
+		return record.Record{}, cerrors.Errorf("error unwrapping metadata: %w", err)
+	}
+
+	return record.Record{
+		Key:       rec.Key,
+		Position:  rec.Position,
+		Operation: operation,
+		Payload: record.Change{
+			Before: before,
+			After:  after,
+		},
+		Metadata: metadata,
+	}, nil
+}
+
+func (d *DebeziumUnWrapper) validateRecord(data record.StructuredData) error {
+	var multiErr error
+	if _, ok := data[DebeziumFieldBefore]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldBefore))
+	}
+	if _, ok := data[DebeziumFieldAfter]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldAfter))
+	}
+	if _, ok := data[DebeziumFieldSource]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldSource))
+	}
+	if _, ok := data[DebeziumFieldOp]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldOp))
+	}
+	// ts_ms and transaction can be empty
+	return multiErr
+}
+
+func (d *DebeziumUnWrapper) unwrapMetadata(rec record.Record, payload record.StructuredData) (record.Metadata, error) {
+	if payload[DebeziumFieldSource] == nil {
+		return nil, nil
+	}
+	meta := record.Metadata{}
+	mp, ok := payload[DebeziumFieldSource].(map[string]interface{})
+	if !ok {
+		return nil, cerrors.Errorf("%q is not formatted as a map", DebeziumFieldSource)
+	}
+	for k, v := range mp {
+		if str, ok := v.(string); ok {
+			meta[k] = str
+		} else {
+			return nil, cerrors.Errorf("the value %q from the field %q is not a string", v, DebeziumFieldSource)
+		}
+	}
+	// append original record's metadata to meta
+	for k, v := range rec.Metadata {
+		meta[k] = v
+	}
```

Comment:
❕If the OpenCDC record and Debezium record contain the same metadata key, the OpenCDC one takes precedence. I would expect the Debezium value to take over tbh, same as the operation is overwritten or payload before.

Reply:
agreed